### ISSUE TITLE
clean calico config at worker start

### ISSUE THREAD

**Issue by calvix** [giantswarm/k8scloudconfig#162]
When for some reason worker is recreated (ie pod delete/scale or w/e reason) calico wont  start because there will already be records with same IP and different machine ID/hostname

we should add `systemd` unit that will clear etcd  

this is how i did that manually:
```
data=`sudo etcdctl --endpoints=https://etcd.d073w.g8s.kc.cg.internal:443 --cert-file /etc/kubernetes/ssl/etcd/client-crt.pem --key-file /etc/kubernetes/ssl/etcd/client-key.pem --ca-file /etc/kubernetes/ssl/etcd/client-ca.pem ls --recursive /calico | grep dkpen`

for i in $data; do sudo etcdctl --endpoints=https://etcd.d073w.g8s.kc.cg.internal:443 --cert-file /etc/kubernetes/ssl/etcd/client-crt.pem --key-file /etc/kubernetes/ssl/etcd/client-key.pem --ca-file /etc/kubernetes/ssl/etcd/client-ca.pem rm $i >/dev/null 2>/dev/null; done;

```
`dkpen` is the worker name taken from pod name `worker-dkpen-3853139292-n0m54`

**Comment by r7vme** [comment]
I've started to work on this task.

After asking in Calico stack they recommened to take look on 

https://github.com/caseydavenport/calico-node-controller

I was able to test it with on-prem guest cluster. Died node were successfully removed from Calico state.

What is missing:
1) We need to unify node names in our kubernetes (Use hostnames everywhere, currently AWS use hostnames, onprem IP addresses). Calico by default uses hostnames.
2) "calico-node-controller" need to be properly packed to Docker/Helm
3) "calico-node-controller" is raw PoC. So i need evaluate it more and work on TODO items that mentioned in README.

**Comment by calvix** [comment]
duplicate https://github.com/giantswarm/k8scloudconfig/issues/162#issuecomment-331082448

**Comment by xh3b4sd** [comment]
@r7vme can you give a status update? AFAIK you worked on this and there have been PRs in the calico node controller fork.

**Comment by r7vme** [comment]
Sure.

- Initial worked version https://github.com/giantswarm/calico-node-controller
- Installed in `gauss` (see kube-system namespace)
- [Manual test plan](https://docs.google.com/document/d/1_KcPdkmxH13A-Z13RdoYcOUiUd6ANCUcAq-IEtlOsUc/edit)

TODO:
- Execute test plan in `gauss`, if ok proceed further
- Create helm manifests
- Create CircleCI stuff
- Release to AWS host clusters (On-prem has a blocker https://github.com/giantswarm/hive/pull/140)
- Evaluate in prod. with host clusters, if OK.
- Release to Guest clusters.

Btw, i want someone from Platform engineers to take look on the code:
- unit tests
- potential improvments (See [todo list](https://github.com/giantswarm/calico-node-controller/blob/master/README.md#todos))

**Comment by xh3b4sd** [comment]
I scanned the code. When we notice a problem we can fix it. Other thoughts are these.

- Why do we even need to run something like this? Why is calico itself not doing this? Like, keeping the nodes in sync and updated is a responsibility of calico itself IMO. We should maybe get in contact with the community here. As a calico user needing to run the node controller additionally does not feel right. WDYT? 
- The cache informers used are the usual open source ones. This is not really good but can be good enough. What we have to test is how good the sync works in the wild. How long does it take to update calico?
- I am not sure why there are multiple (5) goroutines to do the same work. I miss something here.

**Comment by r7vme** [comment]
Thanks for quick feedback. Appreciate

> Why do we even need to run something like this? Why is calico itself not doing this? Like, keeping the nodes in sync and updated is a responsibility of calico itself IMO. We should maybe get in contact with the community here. As a calico user needing to run the node controller additionally does not feel right. WDYT?

This is what i did. I contacted to them in slack. See 
![2017-10-02-12 12 23-screenshot](https://user-images.githubusercontent.com/1277504/31073145-04e15294-a76b-11e7-896e-f9067338f565.png)

> The cache informers used are the usual open source ones. This is not really good but can be good enough. What we have to test is how good the sync works in the wild. How long does it take to update calico?

Just to make sure: We are not talking here about Calico version updates. 

[remove node](https://github.com/giantswarm/calico-node-controller/blob/master/controller.go#L264) takes as any API call really small amount of time, but i'm not sure that you asking about that.

> I am not sure why there are multiple (5) goroutines to do the same work. I miss something here.

i'm not Developer, but for me it looks like there are [multiple workers](https://github.com/giantswarm/calico-node-controller/blob/master/controller.go#L126-L129) consuming events from Queue to be able to process multiple events simulaneously.

**Comment by xh3b4sd** [comment]
> Just to make sure: We are not talking here about Calico version updates.

The calico node controller uses an informer to get events from the API. The informer is configured with the resync period `0`. This means the informer creates a watcher and watches the API, until it breaks. This may happen sooner or later. Until the watcher dies (because the API closes it after some timeout or you close it on your own or the watcher gets disrupted because of network issues etc.) the current nodes are not relisted. So changes within calico are not managed in case no changes happen to the k8s nodes. For the work the controller wants to do this might be good enough but we have to test it. I might miss some details here. 

> i'm not Developer, but for me it looks like there are multiple workers consuming events from Queue to be able to process multiple events simulaneously.

I just wonder why this would be necessary since there shouldn't be that many concurrent node operations. I might miss some details here as well. Not a critical thing though.

**Comment by calvix** [comment]
I have a second opinion on this.
I am not sure it will cover this story, as this means that after worker/master reboot we may still need to go to the cluster and manually delete old Nodes in order to fix this.

So from my point of view, it still makes sense to have custom calico init container which will clear etcd before the start of calico. (by clear i mean remove any entries with same hostname or ip)

**Comment by xh3b4sd** [comment]
> after worker/master reboot we may still need to go to the cluster and manually delete old Nodes and wait until it reconciles in order to fix this

Why is this? This is the task the controller should take care off. This is for instance the reason why the resync period of the cache informer should be quite low. I explained that above already. When the controller does not fix rebooted nodes, we have to fix it in the controller. IMO the controller is the right approach from the design point of view. Having another sidecar would be backwards to me. The controller should just do the proper reconciliation. I can help with this if necessary.

**Comment by calvix** [comment]
To my understanding. It cleans calico configuration for a node when its deleted from k8s.
Calico cant predicts if the `NotReady` not is just offline for few sec or if its dead node and should be removed.

**Comment by xh3b4sd** [comment]
IMO when a node "disappears" it should be deregistered and when it "appears" it should be registered. This is why the resync period should be quite low because it constantly checks the current and the desired state and reconciles towards it. The controller can just do this. It maybe does not do it reliably right now but this is what I offered my help for.

**Comment by calvix** [comment]
if a machine dies how it will deregister? I can see deregister when machine is properly shutdown, but that doesn't happen always

**Comment by xh3b4sd** [comment]
The controller should keep state of the existing machines and check if all the machines it holds state for are there. In case one is missing it sees it by comparing the state. Then it can deregister.

**Comment by calvix** [comment]
I think I am missing something since I don't see how would calico controller do this.
I am going to leave it to you guys.

**Comment by xh3b4sd** [comment]
Easy, FYI the controller watches the nodes within the k8s API and it has access to the calico cache/store IIRC. This is then totally doable. I am just not sure it does the right thing reliably as it is right now, which is why I asked for serious testing.

**Comment by calvix** [comment]
But it can detect that machine is down because it was deleted or machine is down because there is some small network hickup?

**Comment by xh3b4sd** [comment]
It should track its own internal state about the available nodes. When a node is deleted in k8s and thus not visible in the k8s API but present in the internal state, it has to be deleted in the calico cache/store.

**Comment by calvix** [comment]
`When a node is deleted in k8s ` but this does not happen automatically when kvm pod is recreated.

**Comment by r7vme** [comment]
Exactly, we need to take care about cleaning up physically deleted nodes.

Just went thru kubernetes behaviour in that case. kubernetes takes care about this in case if it works in cloud provider. I can confirm that in case of AWS nodes cleaned up automatically from k8s. Magic happens in [k8s node controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/node_controller.go#L757-L758).

Regarding KVM i don't have immediate idea how to do that. Right way probably it to write own node-controller which will sync with kvm-operator (some source of truth) and makes sure to delete node.

So yes, calico-node-controller solves issue only for Calico nodes and can not (should not) do clean ups for kubernetes.

**Comment by r7vme** [comment]
Calico-node-controller released to all AWS host clusters.

For guest clusters, i think we are good to go too. But it won't benefit for on-prem guest clusters, because (see comment above) deleted nodes are staying in NotReady state and should be deleted first.

I'll see how calico-node-controller will live in host cluster and after a week or so i'll prepare PR.

**Comment by calvix** [comment]
focus of this issue should be primary for guest cluster so we need to find some solution for it.

**Comment by xh3b4sd** [comment]
I do not understand the problem with the guest clusters. So guest cluster nodes are just `NotReady` and not deleted. This is still a state we can observe and act upon. The controller has to be adapted though.

**Comment by r7vme** [comment]
Let's summarize what we need/have.

Initial problem was that after node recreation new node were not able to join, because Calico had old node with the same ip address.

"vanilla "calico-node-controller" fixes this problem 100% for any AWS based cluster. Because k8s clean ups dead nodes automatically (--cloud-provider=aws) and calico-node-controller replicates this to Calico. PROFIT!

This problem is not fixed for on-prem guest clusters, because k8s does not clean up dead nodes automatically.

### TL;DR to fix this issue we need to fix another issue: **Clean up dead nodes from k8s in on-prem clusters.**

imo this is not calico-node-controller responsibility as it does own job, replicate k8s state to Calico. Looks like we are not alone with this problem and in 1.8 they allow to run cloud specific stuff as separate service [1] . We can write smth like this for kvm-operator. We just need to scan if pod with specific name exists in host cluster and if not delete this node.

Should we create separate issue and close this one as fixed for AWS?

1 - https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/

**Comment by xh3b4sd** [comment]
Isn't there some garbage collection in Kubernetes? Can't we configure nodes with a `NotReady` state to be garbage collected after a couple of minutes? 

I still don't understand why it is important to remove the nodes though. What I understood was that calico has internal state and the node controller cleans this state as soon as some node is detected as "dead". Why is it necessary to completely purge a Kubernetes node here? Will IPs be reused or are there other conflicts? IMO the node controller could just work in a KVM environment, but we have to adapt the detection for "dead" nodes on KVM. I do not expect this change to be huge.

**Comment by r7vme** [comment]
>  What I understood was that calico has internal state and the node controller cleans this state as soon as some node is detected as "dead". 

When node deleted this info should be proposed in following way.

Source of truth (e.g AWS EC2 instance list) ==1==> K8s ==2==> Calico.

Step 1 covered by k8s natively for clouds. Not covered for on-prem.
Step 2 covered by calico-node-controller for all.

Problem that we don't know source of truth with baremetal.

Yes, i'm also thinking about purging NotReady nodes. But it's not the best thing to do, e.g you have network issues with master, so it decides that all nodes are NotReady and deletes them, when they come back, there will be mess.

Can we somehow supply master with list of worker pods? e.g. Have service that will have access to host cluster and RBAC restriction for read-only access to specific namespace. Then this controller (or even properly implemented cloud-controller `--cloud-provider=external`) can just list pods and do the same thing that kubernetes does for AWS. See the [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L143-L145).

But as a starting point we can implement simple (wrong logic):
1) Get all nodes in NotReady state
2) Check network connectivity to node (e.g. icmp) <--- i would replace this with pod exists :/
3) Delete if node unreachable

**Comment by r7vme** [comment]
@xh3b4sd what is your ideas how we can get info about pods from host cluster/kvm-operator?

**Comment by xh3b4sd** [comment]
So our source of truth is if the guest cluster pod which runs the kubelet exists or not. The node controller is already running on the host cluster AFAIK so fetching such information should work. It is only that we have to adapt the node controller to support our Giant Swarm specific stuff but this can be implemented in a abstracted/general way. There can be some onprem provider that supports checking for pods or something.

**Comment by r7vme** [comment]
> The node controller is already running on the host cluster AFAIK so fetching such information should work.

what node controller are referring to? calico-node-controller or node controller that sits inside kube-contoller-manager?

**Comment by xh3b4sd** [comment]
The calico one.

**Comment by r7vme** [comment]
It's not running on on-prem host cluster, because we are using ip addresses there, but calico uses host names. https://github.com/giantswarm/hive/pull/140

**Comment by calvix** [comment]
but calico-node-controller is running and managing k8s where it runs. It cannot also manage all the guest clusters

**Comment by xh3b4sd** [comment]
I do not feel like I help you guys here with this. Go ahead on your own behalf. Let me know when I can help somehow. At this point I have the feeling I only confuse and make the story take longer. In my head I think I know how to solve it but I probably miss some details you are aware of and I don't. So I shut up now. :D

**Comment by r7vme** [comment]
Okay, let me take this an as a starting point implement `kvm-operator-node-controller` that will run inside cluster and just clean up nodes on NotReady state after timeout. I'll keep in mind that later we will integrate real "source of truth" there.

It will be simple Go service (script) that will grab k8s nodes and if needed execute node delete.

@calvix WDYT?

**Comment by calvix** [comment]
sounds good to me, would be also good idea to try ping node if its really dead before deleting

**Comment by calvix** [comment]
it could also run outside in host cluster and iterate on all `kvm` TPR and use certs that cert-operator created

This way we are not  screwed when this pods cant run in  the  guest cluster because of bad nodes

**Comment by r7vme** [comment]
Hello,

I'm doing some progress in kvm-operator-node-controller,  but not ready to commit yet.

At the moment i'm internally modifying this [PR](https://github.com/giantswarm/kvm-operator-node-controller/pull/1) to run in host cluster.

- implemented provider that checks if pod exists and Ready
- added kubernetes files
- need to finalize some code things (e.g. micrologger and set namespace explicitly)

Current blocker:
- cert-operator does not support creating certificates in namespaces (only in default) - i'll probably try to tackle that on my own.

**Comment by r7vme** [comment]
Good evening, i did some progress with node-controller and ready for review. Tested in lycan. Tomorrow i also want to write down test plan and go thru one more time.

What we have: node-controller (for k8s) runs in host cluster in guest cluster namespace as pod. It will makes sure to clean up node from guest cluster if it was deleted or restarted. Certificates for guest cluster are taken from api secret. From kvm-operator side we have only 1 additional deployment resource.

All PRs (all tested together):
https://github.com/giantswarm/kvm-operator-node-controller/pull/4
https://github.com/giantswarm/kubernetesd/pull/289
https://github.com/giantswarm/kvm-operator/pull/221
https://github.com/giantswarm/kvmtpr/pull/51

**Comment by xh3b4sd** [comment]
> runs in host cluster in guest cluster namespace as pod

So in host or in guest? Based on the KVM operator changes it is in guest.

**Comment by r7vme** [comment]
pod in host cluster. pod namespace is XXXXX. Where XXXXX is id of guest cluster :D

**Comment by r7vme** [comment]
I'm waiting on one PR approval.

All PRs (all tested together):
https://github.com/giantswarm/kvm-operator-node-controller/pull/4 - in review
https://github.com/giantswarm/kubernetesd/pull/289 - approved
https://github.com/giantswarm/kvm-operator/pull/221 - approved
https://github.com/giantswarm/kvmtpr/pull/51 - merged

When all above will be merged, we still need to wait when new version of k8scloudconfig will be vendored to kvm-operator (it's delayed because of RBAC support). Then this issue can be closed.

**Comment by calvix** [comment]
Closing.
